### PR TITLE
Add cache_from and cache_to support for multiplatform builds

### PR DESCRIPTION
**Pull request by shanejbrown**
Add cache_from and cache_to support for multiplatform builds. Passes cache_from and cache_to arguments to python_on_whales.

### What does this PR do?

Improve multiplatform support.

### Merge requirements satisfied?
- [x] I have updated the documentation or no documentation changes are required.
- [x] I have added tests to cover my changes.
- [x] I have updated the base version in ``setup.py`` (if appropriate).

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -250,16 +250,40 @@ shows the different configuration options available:
...
 
+        # The following applies to multiplatform builds.
+        # Specify Docker images to consider as cache sources,
+        # similar to the --cache-from option used by Docker.
+        # cache_from: Works only with the container driver. Loads the cache
+        #     (if needed) from a registry `cache_from="user/app:cache"`  or
+        #     a directory on the client `cache_from="type=local,src=path/to/dir"`.
+        #     It's also possible to use a dict or list of dict form for this
+        #     argument. e.g.
+        #     `cache_from=dict(type="local", src="path/to/dir")`
```
Comment (bluesliverx) [inline README.rst:271]: You can use `dict` python code?

File: README.rst [hunk 2]
```diff
@@ -250,16 +250,40 @@ shows the different configuration options available:
...
+        # cache_to: Works only with the container driver. Sends the resulting
+        #     docker cache either to a registry `cache_to="user/app:cache"`,
+        #     or to a local directory `cache_to="type=local,dest=path/to/dir"`.
+        #     It's also possible to use a dict form for this argument. e.g.
+        #     `cache_to=dict(type="local", dest="path/to/dir", mode="max")`
+        cache_from:
+          my-images/image:PR-123
```
Comment (bluesliverx) [inline README.rst:278]: This format looks strange to me, shouldn't the string be on the previous line?

File: README.rst [hunk 3]
```diff
@@ -250,16 +250,40 @@ shows the different configuration options available:
...
+        <or>
+        cache_from:
+          - {"type": "local", "src": "path/to/dir"}
```
Comment (bluesliverx) [inline README.rst:281]: This is mixing JSON with YAML, maybe just do YAML?

```
cache_from:
- type: local
  src: path/to/dir
```

File: README.rst [hunk 4]
```diff
@@ -250,16 +250,40 @@ shows the different configuration options available:
...
+
+        cache_to:
+          {"type": "local", "dest": "my-images/image:PR-123", "mode": "max"}
```
Comment (bluesliverx) [inline README.rst:284]: using json again, maybe just use pure yaml

File: buildrunner/__init__.py [hunk 5]
```diff
@@ -365,10 +365,31 @@ def run(self):  # pylint: disable=too-many-statements,too-many-branches,too-many
                     step_name,
                     step_config,
                 ) in buildrunner_config.run_config.steps.items():
+                    # Reset the cache_from and cache_to for each step to the global values
+                    multi_platform._cache_from = (
```
Comment (bluesliverx) [inline buildrunner/__init__.py:369]: Accessing private variables is an anti-pattern, maybe you can use a method instead?

File: buildrunner/config/validation.py [hunk 6]
```diff
@@ -253,9 +256,13 @@ def validate_multiplatform_build(
...
 
-            if step.build.cache_from:
+            if (
+                step.build.platforms
+                and isinstance(step.build.cache_from, list)
+                and all(isinstance(x, str) for x in step.build.cache_from)
+            ):
                 raise ValueError(
-                    f"cache_from is not allowed in multi-platform build step {step_name}"
+                    f"{BUILD_MP_CACHE_ERROR_MESSAGE} {step_name} cannot be a list(str) for multiplatform images {type(step.build.cache_from)}]"
```
Comment (bluesliverx) [inline buildrunner/config/validation.py:265]: The last `type` call here will just show something nonsensical to the user, can you can add parenthesis or something to make it clear it's for debug purposes?

File: README.rst (hunk 2 above)
Comment (shanejbrown) [reply README.rst:278]: fixed

File: README.rst (hunk 3 above)
Comment (shanejbrown) [reply README.rst:281]: fixed

File: README.rst (hunk 4 above)
Comment (shanejbrown) [reply README.rst:284]: fixed

File: README.rst (hunk 1 above)
Comment (shanejbrown) [reply README.rst:271]: Actually `cache_from=dict(type="local", src="path/to/dir")` is valid python code. But I already changed it to use {}

File: buildrunner/__init__.py (hunk 5 above)
Comment (shanejbrown) [reply buildrunner/__init__.py]: Added set methods

File: buildrunner/config/validation.py (hunk 6 above)
Comment (shanejbrown) [reply buildrunner/config/validation.py]: Removed that last part of the message.

File: buildrunner/steprunner/tasks/build.py [hunk 7]
```diff
@@ -54,21 +54,25 @@ def __init__(
...
-                        cache_from_image
-                    )
-                    self.step_runner.log.write(
-                        f"Using cache_from image: {cache_from_image}\n"
-                    )
-                except Exception:  # pylint: disable=broad-except
-                    self.step_runner.log.write(
-                        f"WARNING: Unable to pull the cache_from image: {cache_from_image}\n"
-                    )
+                if isinstance(cache_from_image, str):
```
Comment (bluesliverx) [inline buildrunner/steprunner/tasks/build.py:60]: Oh, this is in the single platform code, gotcha.

**Review by bluesliverx — APPROVED** [review]

File: buildrunner/steprunner/tasks/build.py (hunk 7 above)
Comment (shanejbrown) [reply buildrunner/steprunner/tasks/build.py:60]: I could change it to support multiplatform and it should be beneficial for our build nodes.

Comment (shanejbrown) [reply buildrunner/steprunner/tasks/build.py:60]: I'll create a new ticket to track this, so we can test it too.

**Pull request merged**